dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,320 @@
+            with open(
+                self.gdrive_user_credentials_path, "w"
+            ) as credentials_file:
+                credentials_file.write(
+                    os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA)
+                )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        GoogleAuth.DEFAULT_SETTINGS["client_config"] = {
+            "client_id": self.gdrive_client_id,
+            "client_secret": self.gdrive_client_secret,
+            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+            "token_uri": "https://oauth2.googleapis.com/token",
+            "revoke_uri": "https://oauth2.googleapis.com/revoke",
+            "redirect_uri": "",
+        }
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials"] = True
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials_backend"] = "file"
+        GoogleAuth.DEFAULT_SETTINGS[
+            "save_credentials_file"
+        ] = self.gdrive_user_credentials_path
+        GoogleAuth.DEFAULT_SETTINGS["get_refresh_token"] = True
+        GoogleAuth.DEFAULT_SETTINGS["oauth_scope"] = [
+            "https://www.googleapis.com/auth/drive",
+            "https://www.googleapis.com/auth/drive.appdata",
+        ]
+
+        # Pass non existent settings path to force DEFAULT_SETTINGS loading
+        gauth = GoogleAuth(settings_file="")
+        gauth.CommandLineAuth()
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            os.remove(self.gdrive_user_credentials_path)
+
+        gdrive = GoogleDrive(gauth)
+        return gdrive
+
+    def create_drive_item(self, parent_id, title):
+        return gdrive_retry(
+            lambda: self.gdrive_create_folder(title, parent_id)
+        )()
+
+    def get_drive_item(self, name, parents_ids):
+        if not parents_ids:
+            return None
+        query = " or ".join(
+            "'{}' in parents".format(parent_id) for parent_id in parents_ids
+        )
+
+        query += " and trashed=false and title='{}'".format(name)
+
+        item_list = gdrive_retry(lambda: self.gdrive_list_file(query))()
+        return next(iter(item_list), None)
+
+    def resolve_remote_file(self, parents_ids, path_parts, create):
+        for path_part in path_parts:
+            item = self.get_drive_item(path_part, parents_ids)
+            if not item and create:
+                item = self.create_drive_item(parents_ids[0], path_part)
+            elif not item:
+                return None
+            parents_ids = [item["id"]]
+        return item
+
+    def subtract_root_path(self, parts):
+        if not hasattr(self, "root_id"):
+            return parts, [self.path_info.bucket]
+
+        for part in self.path_info.path.split("/"):
+            if parts and parts[0] == part:
+                parts.pop(0)
+            else:
+                break
+        return parts, [self.root_id]
+
+    def get_path_id_from_cache(self, path_info):
```

Comment:
This cache code is hard to understand. Too many entry points, too many moving parts.

Can we limit it to two entry points `.path_info_to_ids()` and `.id_to_path_info()` and two cache structures backing them, which will be updated in sync? Then store not only first level dirs there, but all of them, including root dir. Then we can make them all lazily, while still using cache.

Implementation considerations:
- it might be easier to use `.path_to_ids()` and `id_to_path()` or not),
- we still probably want to fill cache for root under the thread lock,
- we only want first id from `.path_info_to_ids()` when we are going to upload a file under it, but we should go over all of the ids on listing
- we might need to store parent-child relations to quickly list first level dirs as it is done now. Not sure this is worth the hustle. Maybe listing all of them each time is ok.

Reply:
> we still probably want to fill cache for root under the thread lock,

@Suor it's my main blocker to implement lazy gdrive init for now.
Filling folders cache requires queries to remote gdrive, but it's under Lock `@wrap_with(threading.Lock())` already and recursive call causes deadlock.

It seems I can't see acceptable workaround which is better than current implementation.

Reply:
use `threading.RLock()`?

Reply:
@casperdcl oh cool, thanks, that should help!

Reply:
lazy gdrive init with `RLock` seems works good, thanks!

Since cache properties are used (and might be used) before gdrive is initialized, I have ended for the moment with ugly trigger to init gdrive https://github.com/iterative/dvc/pull/2551/commits/9f6d4696cf6fd298c4ce25ce550b8ca3f1327312#diff-d5fd1a3af39d64db31b96286f8985dbeR136

I've slightly renamed "get from cache" methods, but agree that it makes sense to simplify implementation.

Reply:
@Suor resolve?

Reply:
Have simplified caching logic with full remote path to cache ( as suggested by @Suor  ) https://github.com/iterative/dvc/pull/2551/commits/9c156c2259d1c8ba25dee931a5aa27162c377ad5

> Can we limit it to two entry points .path_info_to_ids() and .id_to_path_info() and two cache structures backing them, which will be updated in sync? Then store not only first level dirs there, but all of them, including root dir. Then we can make them all lazily, while still using cache.

Since it's not possible to create nested remote path with single API call, current logic tries to search cache for already created sub paths and loops to create missed sub dirs.
Hopefully, it should be fine to go with current revised implementation. Please let me know.

Reply:
It is still looks hard to understand to me. It doesn't matter how many API calls are behind a python call. So why can't we limit the whole thing to those two entry points? 

We shouldn't need all those `cached_dirs` props also.

Reply:
>  So why can't we limit the whole thing to those two entry points?

we can limit into 2 entry points, but my current opinion is that the implementation of these 2 entry points still will not be nice. I can try :)
To bring more clarity why I think it will be not nice: since DVC works with usual paths ( "dir1/dir2/dir3/file" ), it's not compatible with Google Drive which uses `id`s as unique identifier of the file or folder (and to get `id` of `dir2` we need to get `id` of `dir1` first and send new request to search for child of `parent_id` with title `dir2` ).
It means that we can't cache entry points inputs/outputs directly, because it doesn't dive us any optimization.
Well, it seems it will be needed to have additional helper method which will take 2 params - `parent_id`, `title`  and return the actual `id`. Such helper method will replace `cached_dirs` props looks like.

Reply:
Moved as To-Do task as part of https://github.com/iterative/dvc/issues/2865